### PR TITLE
blockchain: Add chaingen harness AdvanceToHeight.

### PR DESCRIPTION
**Pull request by dnldd**
This adds `AdvanceToHeight` to the `chaingen` harness to allow advancing the chain to an arbitrary height while controlling the number of tickets purchased per block after the first coinbase maturity.

NOTE: This PR was originally intended to add a chain continuity check, but it was modified to only introduce the aforementioned logic in favor of #2398.

### CODE REVIEW

File: blockchain/difficulty.go [hunk 1]
```diff
@@ -540,6 +540,69 @@ func (b *BlockChain) sumPurchasedTickets(startNode *blockNode, numToSum int64) i
 	return numPurchased
 }
 
+// immatureTickets returns the total number of immature tickets at the
+// current tip and a map of immature ticket counts keyed by the block height at
+// which they were bought at.
+func (b *BlockChain) immatureTickets() (int64, map[int64]int64, error) {
+	tip := b.bestChain.Tip()
+	rangeStart := tip.height - int64(b.chainParams.TicketMaturity)
+	startNode := b.bestChain.nodeByHeight(rangeStart)
```
Comment (sefbkn) [inline blockchain/difficulty.go]: Since this walks the chain through `node.parent`, should we instead start iterating at the newest block in the range rather than the oldest?

Comment (dnldd) [reply blockchain/difficulty.go]: You're right, good catch.

File: blockchain/difficulty.go [hunk 2]
```diff
@@ -540,6 +540,58 @@ func (b *BlockChain) sumPurchasedTickets(startNode *blockNode, numToSum int64) i
...
 
+// immatureTickets returns the total number of immature tickets at the
+// current tip and a map of immature ticket counts keyed by the block height at
+// which they were bought at.
+func (b *BlockChain) immatureTickets() (int64, map[int64]int64) {
+	tip := b.bestChain.Tip()
+	var numImmature int64
+	immatureSet := make(map[int64]int64, b.chainParams.TicketMaturity)
+	for node, numTraversed := tip, int64(0); node != nil &&
+		numTraversed <= int64(b.chainParams.TicketMaturity); numTraversed++ {
```
Comment (sefbkn) [inline blockchain/difficulty.go]: Shouldn't this be `<` to avoid double-counting mature tickets?

To illustrate where I'm coming from, let's assume `TicketMaturity = 1` and `tip` has a height of 10. A ticket included in block 9 would mature in block 10 and therefore would not be immature. The nuance here is that despite block 9's ticket purchases being mature in block 10 the following constraints are in play that prevent it from voting until block 11:

* a utxo must precede its redeemer
* votes must precede tickets in a block
* the lock time relative to the block the tx was mined in must elapse before it is spendable (TicketMaturity)

Regardless if there is an off-by-one scenario here or not, there is room for a comment explaining what's going on and why for future devs since this part of the code is really tricky due to the non-obvious relation between transaction ordering in a block and maturity requirements defined elsewhere.

File: mining.go [hunk 3]
```diff
@@ -958,6 +958,18 @@ func (g *BlkTmplGenerator) NewBlockTemplate(payToAddress dcrutil.Address) (*Bloc
...
+		tickets := g.chain.LiveTicketsAtNextMaturity()
+		if tickets < uint64((g.chainParams.TicketsPerBlock/2)+1) {
+			futureHeight := best.Height + int64(g.chainParams.TicketMaturity) + 1
+			msg := fmt.Sprintf("chain continuity not guaranteed after "+
+				"height %d, remaining live tickets after ticket maturity "+
+				"period from current tip are: %d)", futureHeight, tickets)
+			minrLog.Debug(msg)
+			return nil, fmt.Errorf(msg)
+		}
+
```
Comment (sefbkn) [inline mining.go]: This seems only to be reactive rather than both proactive and reactive since it doesn't take the block template's ticket purchases into account. To not waste work on a dead-end chain is definitely part of the solution, but manual intervention to reorg the chain would still be required in a worst-case scenario. I think the warning signs for the ticket pool exhaustion should look around `TicketMaturity` blocks into the future and assume that all blocks would have 5 votes at least.

It seems like this problem should ultimately be solved by the consensus layer, and the objective of this PR is not clear as-is.

File: blockchain/difficulty.go [hunk 4]
```diff
@@ -540,6 +540,58 @@ func (b *BlockChain) sumPurchasedTickets(startNode *blockNode, numToSum int64) i
...
+}
+
+// LiveTicketsAtNextMaturity returns the total number of live tickets
+// available after the ticket maturity period from the current tip.
+// This assumes all blocks mined during the period are voted on
+// by the maximum number of tickets allowed per block.
+func (b *BlockChain) LiveTicketsAtNextMaturity() uint64 {
+	tip := b.bestChain.Tip()
+	b.chainLock.RLock()
+	liveTickets := int64(len(tip.stakeNode.LiveTickets()))
```
Comment (sefbkn) [inline blockchain/difficulty.go]: `PoolSize()` appears to give the same result in less operations, though requesting tickets for a new block template is pretty infrequent.

File: blockchain/difficulty.go (hunk 2 above)
Comment (dnldd) [reply blockchain/difficulty.go]: Tickets mature at `TicketMaturity` but aren't immediately spendable, they become spendable at `TicketMaturity+1`, referring to the example provided the ticket included in block 9 would mature in block 10 and be available to vote in block 11. 

Thanks for pointing this out, will update with comments.

File: mining.go (hunk 3 above)
Comment (dnldd) [reply mining.go]: Yes, the check ultimately should added at the consensus layer. However this implementation at the template generation level prevents to some extent those long reorgs that have occurred on testnet. You can think of this as a stop-gap solution until it's implemented at consensus level.

File: blockchain/difficulty.go [hunk 5]
```diff
@@ -540,6 +540,34 @@ func (b *BlockChain) sumPurchasedTickets(startNode *blockNode, numToSum int64) i
...
+// LiveTicketsAtNextMaturity returns the total number of live tickets
+// available after the ticket maturity period from the provided block.
+// This assumes all blocks mined during the period are voted on
+// by the maximum number of tickets allowed per block.
+func (b *BlockChain) LiveTicketsAtNextMaturity(hash *chainhash.Hash) uint64 {
+	b.chainLock.Lock()
+	defer b.chainLock.Unlock()
+
+	node := b.index.lookupNode(hash)
+	liveTickets := int64(node.stakeNode.PoolSize())
```
Comment (davecgh) [inline blockchain/difficulty.go:552]: This is the only call that needs the lock.  The rest are accessing immutable fields.

Also this will crash if called with a hash that doesn't exist and it isn't checking if the data is available to properly support pruning.  I suggest taking a look at something like `BlockByHash` for an example.

e.g.

```Go
	node := b.index.LookupNode(hash)
	if node == nil || !b.index.NodeStatus(node).HaveData() {
		return nil, fmt.Errorf("block %s is not known", hash)
	}
	
	b.chainLock.Lock()
	liveTickets := int64(node.stakeNode.PoolSize())
	b.chainLock.Unlock()

	...
```

File: blockchain/difficulty.go [hunk 6]
```diff
@@ -540,6 +540,34 @@ func (b *BlockChain) sumPurchasedTickets(startNode *blockNode, numToSum int64) i
...
+
+	// Iterate through all nodes in the last ticket maturity period to
+	// get the number of tickets maturing in the next ticket maturity period.
+	for ; node != nil && node.height >= rangeStart; node = node.parent {
+		liveTickets += int64(node.freshStake)
+	}
+
+	// Subtract the total number of live tickets needed to progress the
+	// chain to the end of the ticket maturity period from the current tip.
+	liveTickets -= maturityWindow * int64(b.chainParams.TicketsPerBlock)
```
Comment (davecgh) [inline blockchain/difficulty.go:564]: This logic is incorrect prior to SVH.  There are no votes being consumed.  It's important to count this correctly, otherwise the check won't work properly pre SVH.

File: blockchain/difficulty.go [hunk 7]
```diff
@@ -540,6 +540,34 @@ func (b *BlockChain) sumPurchasedTickets(startNode *blockNode, numToSum int64) i
 	return numPurchased
 }
 
+// LiveTicketsAtNextMaturity returns the total number of live tickets
+// available after the ticket maturity period from the provided block.
+// This assumes all blocks mined during the period are voted on
+// by the maximum number of tickets allowed per block.
+func (b *BlockChain) LiveTicketsAtNextMaturity(hash *chainhash.Hash) uint64 {
```
Comment (davecgh) [inline blockchain/difficulty.go:547]: Given the importance of this function, this needs a unit test to hit the edge conditions.  It can create a fake chain like several of the existing tests for this purpose since it just needs to work with nodes.

File: blockchain/difficulty_test.go [hunk 8]
```diff
@@ -1150,3 +1150,260 @@ func TestMinDifficultyReduction(t *testing.T) {
 		}
 	}
 }
+
+// TestLiveTicketsAtNextMaturity ensures the function that ensures the chain
+// always has enough tickets to progress works as expected.
+func TestLiveTicketsAtNextMaturity(t *testing.T) {
```
Comment (rstaudt2) [inline blockchain/difficulty_test.go:1156]: Typo in the commit message for this change:

> This adds tests **covering** LiveTicketsAtNextMaturity and
> its associated edge **cases** to ensure it works as expected.

File: blockchain/difficulty_test.go [hunk 9]
```diff
@@ -1150,3 +1150,260 @@ func TestMinDifficultyReduction(t *testing.T) {
...
+
+		g, teardown := newChaingenHarness(t, params, "chaincontinuitytest")
+		return params, g, teardown
+	}
+
+	params, g, teardown := setup()
+	stakeValidationHeight := uint32(params.StakeValidationHeight)
+	initialContinuityCheckHeight := stakeValidationHeight - 2
+
+	// Progress the chain to two block before StakeValidationHeight, where
```
Comment (rstaudt2) [inline blockchain/difficulty_test.go:1172]:
```suggestion
	// Progress the chain to two blocks before StakeValidationHeight, where
```

File: blockchain/difficulty_test.go [hunk 10]
```diff
@@ -1150,3 +1150,260 @@ func TestMinDifficultyReduction(t *testing.T) {
...
+	// tickets are required to generate voting lottery winners without
+	// buying tickets.
+	g.AdvanceToHeight(initialContinuityCheckHeight, 0)
+
+	// Ensure the the continuity check fails because there a no live tickets.
```
Comment (rstaudt2) [inline blockchain/difficulty_test.go:1177]:
```suggestion
	// Ensure the continuity check fails because there are no live tickets.
```

File: blockchain/difficulty_test.go [hunk 11]
```diff
@@ -1150,3 +1150,260 @@ func TestMinDifficultyReduction(t *testing.T) {
...
+
+	_, g, teardown = setup()
+	ticketsPerBlock := uint32(params.TicketsPerBlock)
+
+	// Progress the chain to two blocks before StakeValidationHeight, where
+	// tickets are required to generate voting lottery winners, buying
+	// ticketsPerBlock each block.
+	g.AdvanceToHeight(initialContinuityCheckHeight, ticketsPerBlock)
+
+	// Ensure the chain continuity check passes because the are enough live
```
Comment (rstaudt2) [inline blockchain/difficulty_test.go:1203]:
```suggestion
	// Ensure the chain continuity check passes because there are enough live
```

File: blockchain/difficulty_test.go [hunk 12]
```diff
@@ -1150,3 +1150,260 @@ func TestMinDifficultyReduction(t *testing.T) {
...
+		t.Fatalf("expected zero live tickets at height #%d, got %d",
+			tipHeight, liveTickets)
+	}
+	if err == nil {
+		teardown()
+		t.Fatalf("expected a continuity error at height #%d", tipHeight)
+	}
+
+	teardown()
+}
```
Comment (rstaudt2) [inline blockchain/difficulty_test.go:1409]: There are a couple of code paths not being tested if you want to hit 100% coverage for `LiveTicketsAtNextMaturity`:

- https://github.com/decred/dcrd/blob/b1f2c12bbdcab81cd40b9b0f51f3a0c028a3a44e/blockchain/difficulty.go#L552
- https://github.com/decred/dcrd/blob/b1f2c12bbdcab81cd40b9b0f51f3a0c028a3a44e/blockchain/difficulty.go#L563
- https://github.com/decred/dcrd/blob/b1f2c12bbdcab81cd40b9b0f51f3a0c028a3a44e/blockchain/difficulty.go#L596-L599

**Review by rstaudt2 — COMMENTED** [review]
Still need to do more review/testing on the last two commits, but had some initial minor comments while reading through.

**Review by davecgh — CHANGES_REQUESTED** [review]
I rebased this to resolve the merge conflicts, but the logic is not quite correct.  It's trivial to see on simnet that you hit the point where you get the following (as expected):

> 2020-10-01 17:57:00.508 [DBG] MINR: chain cannot progress to stake validation height 144, live tickets (0) is below minimum required (3)

However, it doesn't recognize when you have enough ticket purchases in the template to allow you to recover.  I'll work on correcting the logic and addressing the review feedback later tonight.

**Comment by davecgh** [comment]
I spend a bit of timing working on this tonight and found several more things that are incorrect in the logic.  I think I'm just going to introduce the `AdvanceToHeight` (with suggested comments addresses and some other things I saw) and implement the continuity check and tests separately.

**Comment by davecgh** [comment]
I reworked this to only introduce `AdvanceToHeight` to the chaingen harness since it had several show stopping issues such as:

a) The min required tickets were incorrect because if the pool size falls below 5, the lottery funcs fail.
b) It wasn't behaving properly prior to SVH at all and you could easily still create an unrecoverable chain.
c) It wasn't considering things from the PoV of extending the chain with a block with a given number of votes, which means you couldn't properly take into account what will be included in the template from `mempool`.

It also some other less severe issues that could have been addressed, such as improper mutex locking, but due to the above, I just decided to go a different direction altogether and implement a variant that addresses all of the issues as well as can potentially be used in consensus code if gated behind a vote in the future.

**Review by davecgh — APPROVED** [review]
